Created new Shadcn tip (#285)
astrolicious/astro-tips.dev: src/content/docs/tips/shadcn.mdx

---
title: Setting up Shadcn UI
---

import { Tabs, TabItem, Aside, Steps,LinkCard } from "@astrojs/starlight/components";
import ExampleLinks from '~/components/ExampleLinks.astro'

> 

Reviewer: I think it would be good to explain the problem we are solving in this tip, i.e what is layed out in the working with interactive components section 

a quick explainer about the islands being seperate and needing to create our own react component that wraps all the shadcn components parts

## Examples

If you want to see a more interactive example that might help explain or let you test things out better you can: 

<ExampleLinks 
  stackblitzLink="https://stackblitz.com/~/github.com/hkbertoson/astro-shadcn-ui"
  githubLink="https://github.com/hkbertoson/astro-shadcn-ui"
/>

<Steps>
1. Install react using your preferred package manager:
    <Tabs syncKey="shadcn">
      <TabItem label="npm" icon="seti:npm">
        ```sh
        npx astro add react
        ```
      </TabItem>
      <TabItem label="pnpm" icon="pnpm">
        ```sh
        pnpx astro add react
        ```
      </TabItem>
      <TabItem label="yarn" icon="seti:yarn">
        ```sh
        npx astro add react
        ```
      </TabItem>

>       <TabItem label="bun" icon="bun">

Reviewer 2: Do we show bun in other examples? If not let's remove it

Reviewer 3: Separately, we could use https://github.com/HiDeoo/starlight-package-managers

        bunx shadcn@latest init
        ```
      </TabItem>
    </Tabs>
10. That's it, you're all set up to use Shadcn UI in your project!
    You can now start adding components from the Shadcn UI library to your project.
    
        <Tabs syncKey="shadcn">
      <TabItem label="npm" icon="seti:npm">
        ```sh
       npx shadcn@latest add button
        ```
      </TabItem>
      <TabItem label="pnpm" icon="pnpm">
        ```sh
        pnpx shadcn@latest add button
        ```
      </TabItem>
      <TabItem label="yarn" icon="seti:yarn">
        ```sh
        npx shadcn@latest add button
        ```
      </TabItem>
      <TabItem label="bun" icon="bun">
        ```sh
        bunx shadcn@latest add button
        ```
      </TabItem>
    </Tabs>

        The command above will add the `Button` component to your project. You can then import it like this:

        ```astro title="src/pages/index.astro" {2,10}
        ---
        import { Button } from "@/components/ui/button"
        ---

        <html lang="en">
        <head>
            <title>Astro</title>
        </head>
        <body>
            <Button>Hello World</Button>
        </body>
        </html>
        ```

> </Steps>

Reviewer: This section essentially duplicates the [shadcn astro docs](https://ui.shadcn.com/docs/installation/astro), I think in order to keep this tip on topic we can just say follow that getting started

Reviewer 2: Yes I agree with you. I think this tip should
1. point to an example
2. point to the official docs
3. explain tht common mistakes, eg. react context not being shared between islands hence the need to move stuff to its dedicated island

        return (
            <Drawer>
                <DrawerTrigger>Open</DrawerTrigger>
                <DrawerContent>
                    <DrawerHeader>
                        <DrawerTitle>Are you absolutely sure?</DrawerTitle>
                        <DrawerDescription>This action cannot be undone.</DrawerDescription>
                    </DrawerHeader>
                    <DrawerFooter>
                        <Button>Submit</Button>
                        <DrawerClose>
                            <Button variant="outline">Cancel</Button>
                        </DrawerClose>
                    </DrawerFooter>
                </DrawerContent>
            </Drawer>
        );
    };
    ```
3. Import the `MyDrawer` component in the `src/pages/index.astro` file:

    ```astro title="src/pages/index.astro" {2,10}
    ---
    import { Button } from "@/components/ui/button"
    import { MyDrawer } from "@/components/MyDrawer"
    import '@/styles/globals.css'
    ---

    <html lang="en">
    <head>
        <title>Astro</title>
    </head>
    <body>
        <Button>Hello World</Button>
        <MyDrawer client:load />
    </body>
    </html>
    ```
</Steps>

> To********** ******learn more about Shadcn UI, check their official site or their docs:

Reviewer: ```suggestion
To learn more about Shadcn UI, check their official site or their docs:
```

---
title: Setting up Shadcn UI
---

import { Steps, LinkCard } from "@astrojs/starlight/components";
import AllPackageManagers from '../../../components/AllPackageManagers.astro'
import ExampleLinks from '~/components/ExampleLinks.astro'

> When working with Astro and ShadCN UI in a project that involves islands architecture, where interactive components in React need to be wrapped individually in `.tsx` files. This is necessary because, with Astro's islands architecture, React context (or state) is not shared across different islands. Each "island" (or separate component) runs in its own isolated environment, so to manage state or context across interactive components, you need to ensure they are self-contained.

Reviewer 2: ```suggestion
When using Shadcn UI within an Astro project, there is a common mistake that prevents Shadcn UI from working properly. With Astro's [islands architecture](https://docs.astro.build/en/concepts/islands/), React [context is not shared](https://github.com/withastro/roadmap/discussions/742) between islands. That means that calling Shacdn UI components directly in a `.astro` file will fail because islands are not able to communicate together. This can be easily solved by moving those components calls to a `.jsx/tsx` file and importing this one instead.
```

## Examples

If you want to see a more interactive example that might help explain or let you test things out better you can: 

<ExampleLinks 
  stackblitzLink="https://stackblitz.com/github/astrolicious/astro-tips.dev/tree/main/examples/shadcn-ui"
  githubLink="https://github.com/astrolicious/astro-tips.dev/tree/main/examples/shadcn-ui"
/>

## Working with Interactive Components

<Steps>

1. Follow the getting started guide from the Shadcn UI documentation to set up the library in your project.

    <LinkCard

>         title="Shadcn UI Documentation"

Reviewer: ```suggestion
        title="Shadcn Astro Installation Guide"
```

        DrawerTrigger,
    } from "@/components/ui/drawer";

    export const MyDrawer = () => {
        return (
            <Drawer>
                <DrawerTrigger>Open</DrawerTrigger>
                <DrawerContent>
                    <DrawerHeader>
                        <DrawerTitle>Are you absolutely sure?</DrawerTitle>
                        <DrawerDescription>This action cannot be undone.</DrawerDescription>
                    </DrawerHeader>
                    <DrawerFooter>
                        <Button>Submit</Button>
                        <DrawerClose>
                            <Button variant="outline">Cancel</Button>
                        </DrawerClose>
                    </DrawerFooter>
                </DrawerContent>
            </Drawer>
        );
    };
    ```
3. Import the `MyDrawer` component in the `src/pages/index.astro` file:

    ```astro title="src/pages/index.astro" {2,10}
    ---
    import { MyDrawer } from "@/components/MyDrawer"
    import '@/styles/globals.css'
    ---

    <html lang="en">
    <head>
        <title>Astro</title>
    </head>
    <body>
        <MyDrawer client:load />
    </body>
    </html>
    ```
</Steps>

> To learn more about Shadcn UI, check their official site or their docs:

Reviewer: ```suggestion
To learn more about Shadcn UI:
```

<LinkCard

>     title="Shadcn UI Documentation"

Reviewer: ```suggestion
    title="Shadcn Official Docs"
```
